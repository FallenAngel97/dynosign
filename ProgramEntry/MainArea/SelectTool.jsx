import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { draw } from './drawMarchingAnts';

/**
 * Enables the selection box (marching ants) to be shown
 * @module SelectTool
 */

export class SelectTool extends React.Component {
  constructor (props) {
    super(props);
    this.prevX = 0;
    this.currX = 0;
    this.prevX = 0;
    this.prevY = 0;
    this.displaySizes = false;
    this.findxy = this.findxy.bind(this);
  }
  findxy (res, e) {
    if (this.props.changeMouseType.mouseType !== 'select') return;
    if (res === 'down') {
      this.currX = (e.clientX - this.canvas.offsetLeft) * window.devicePixelRatio;
      this.currY = (e.clientY - this.canvas.offsetTop) * window.devicePixelRatio;
      this.prevX = this.currX;
      this.prevY = this.currY;
      this.displaySizes = true;
    }
    if (res === 'up' || res === 'out') { // hide size tooltip on mouse release
      this.displaySizes = false;
      if(this.currX == this.prevX && this.currY == this.prevY) {
        // if clicked just on the screen without dragging - then don't display marching ants
        let ctx = this.canvas.getContext('2d');
        ctx.clearRect(0, 0, this.canvas.width, this.canvas.height);
        draw(0, this.canvas);
        return;
      } else {
        draw(0, this.canvas, this.displaySizes);
      }
    }
    if (res === 'move' && this.displaySizes) {
      this.currX = (e.clientX - this.canvas.offsetLeft) * window.devicePixelRatio;
      this.currY = (e.clientY - this.canvas.offsetTop) * window.devicePixelRatio;
      const currentCoordinate = { x: this.currX, y: this.currY };
      const prevCoordinate = { x: this.prevX, y: this.prevY };
      draw(0, this.canvas, this.displaySizes, currentCoordinate, prevCoordinate);
    }
  }
  render () {
    return <canvas className='canvaslayer'
      width={this.props.width * window.devicePixelRatio}
      height={this.props.height * window.devicePixelRatio}
      style={{ width: this.props.width, height: this.props.height }}
      onMouseMove={(e) => this.findxy('move', e)}
      onMouseDown={(e) => this.findxy('down', e)}
      onMouseUp={(e) => this.findxy('up', e)}
      onMouseOut={(e) => this.findxy('out', e)}
      ref={(canvas) => { this.canvas = canvas }} />
  }
}

SelectTool.propTypes = {
  width: PropTypes.number,
  height: PropTypes.number,
  changeMouseType: PropTypes.object
}

const mapStateToProps = state => state;

export default connect(mapStateToProps)(SelectTool);
