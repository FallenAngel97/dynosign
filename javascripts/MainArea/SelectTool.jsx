import React from "react";
import {connect} from "react-redux";

class SelectTool extends React.Component {
    constructor(props) {
        super(props);
        this.prevX = 0;
        this.currX = 0;
        this.prevX = 0;
        this.prevY = 0;
        this.flag = false;
        this.findxy = this.findxy.bind(this);
        this.draw = this.draw.bind(this);
        this.timerMarchingAnts = undefined;
    }
    draw(offsetDash) {
        var ctx = this.canvas.getContext('2d');
        ctx.beginPath();
        ctx.clearRect(0, 0, this.canvas.width, this.canvas.height);
        ctx.setLineDash([6]);
        ctx.lineDashOffset = offsetDash;
        const diffX = this.currX - this.prevX;
        const diffY = this.currY - this.prevY;
        ctx.strokeRect(this.currX - diffX, this.currY - diffY, diffX,diffY);
        clearTimeout(this.timerMarchingAnts)
        this.timerMarchingAnts = setTimeout(()=> {
            offsetDash++;
            this.draw(offsetDash);
        }, 100);
    }
    findxy(res, e) {
        if(this.props.changeMouseType.mouseType != "select") return;
        if (res == 'down') {
            this.currX = e.clientX - this.canvas.offsetLeft;
            this.currY = e.clientY - this.canvas.offsetTop;
            this.prevX = this.currX;
            this.prevY = this.currY;
            this.flag = true;
        }
        if (res == 'up' || res == "out") {
            this.flag = false;
        }
        
        if (res == 'move') {
            if (this.flag) {
                this.currX = e.clientX - this.canvas.offsetLeft;
                this.currY = e.clientY - this.canvas.offsetTop;
                this.draw(0);
            }
        }
    }
    render() {
        return <canvas className='canvaslayer'
            width={this.props.width}
            height={this.props.height}
            style={{width: this.props.width, height: this.props.height}}
            onMouseMove={(e)=> this.findxy('move', e)}
            onMouseDown={(e)=> this.findxy('down', e)}
            onMouseUp={(e)=> this.findxy('up', e)}
            onMouseOut={(e)=> this.findxy('out', e)}
            ref={(canvas)=>this.canvas = canvas} />
    }
}

const mapStateToProps = state => state;

export default connect(mapStateToProps)(SelectTool);