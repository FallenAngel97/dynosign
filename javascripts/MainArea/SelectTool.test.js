/* eslint-disable no-unused-vars */
/* global jest describe test expect mount */

import React from 'react'
import { SelectTool } from './SelectTool';

jest.useFakeTimers();

describe('SelectTool test set', () => {
  test('SelectTool renders properly', () => {
    let selectTool = mount(<SelectTool />);
    expect(selectTool).toMatchSnapshot();
    expect(selectTool.instance().canvas).toBeTruthy();
  });
  test('Mouse events working correctly', () => {
    const spy = jest.spyOn(SelectTool.prototype, 'findxy');
    const selectTool = mount(<SelectTool changeMouseType={{ mouseType: 'select' }} />);
    selectTool.simulate('mousemove', { preventDefault: () => true })
    expect(spy).toBeCalled();
    selectTool.simulate('mousedown', { preventDefault: () => true })
    expect(spy).toBeCalled();
    selectTool.simulate('mouseup', { preventDefault: () => true })
    expect(spy).toBeCalled();
    selectTool.simulate('mouseout', { preventDefault: () => true })
    expect(spy).toBeCalled();
  });
  test('findxy can handle move', () => {
    const selectTool = mount(<SelectTool changeMouseType={{ mouseType: 'select' }} />);
    const ev = {
      clientX: 200,
      clientY: 200
    };
    selectTool.instance().flag = true;
    selectTool.instance().findxy('move', ev);
    expect(selectTool.instance().currY).toEqual(200);
  });
  test('drawing invoke marching ants', () => {
    const selectTool = mount(<SelectTool changeMouseType={{ mouseType: 'select' }} />);
    selectTool.instance().draw(6);
    expect(setTimeout).toHaveBeenCalledTimes(2);
    expect(setTimeout).toHaveBeenLastCalledWith(expect.any(Function), 100);
  });
});
