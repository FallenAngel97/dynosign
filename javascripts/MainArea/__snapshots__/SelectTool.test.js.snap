// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectTool test set SelectTool renders properly 1`] = `ReactWrapper {}`;
